app/services/onestop_id.rb

```diff
@@ -1,54 +1,114 @@
 require 'addressable/template'
 
-class OnestopId
-  ENTITY_TO_PREFIX = {
-    'stop' => 's',
-    'operator' => 'o',
-    'feed' => 'f',
-    'route' => 'r'
-  }
-  PREFIX_TO_ENTITY = ENTITY_TO_PREFIX.invert
-  PREFIX_TO_MODEL = {
-    's' => Stop,
-    'o' => Operator,
-    'r' => Route,
-    'f' => Feed
-  }
-  MODEL_TO_PREFIX = PREFIX_TO_MODEL.invert
+module OnestopId
+
   COMPONENT_SEPARATOR = '-'
   GEOHASH_FILTER = /[^0123456789bcdefghjkmnpqrstuvwxyz]/
   NAME_TILDE = /[\-\:\&\@\/]/
   NAME_FILTER = /[^a-zA-Z\d\@\~]/
   IDENTIFIER_TEMPLATE = Addressable::Template.new("gtfs://{feed_onestop_id}/{entity_prefix}/{entity_id}")
 
-  attr_accessor :entity_prefix, :geohash, :name
-
-  def initialize(string: nil, entity_prefix: nil, geohash: nil, name: nil)
-    if string && string.length > 0
-      @entity_prefix = string.split(COMPONENT_SEPARATOR)[0]
-      @geohash = string.split(COMPONENT_SEPARATOR)[1]
-      @name = string.split(COMPONENT_SEPARATOR)[2]
-    elsif entity_prefix && geohash && name
-      # Filter geohash and name; validate later
-      @entity_prefix = entity_prefix
-      @geohash = geohash_filter(geohash)
-      @name = name_filter(name)
-    else
-      raise ArgumentError.new('either a string or entity/geohash/name must be specified')
-    end
-    # Check valid OnestopID
-    is_a_valid_onestop_id, errors = OnestopId.validate_onestop_id_string(self.to_s)
-    if !is_a_valid_onestop_id
-      raise ArgumentError.new(errors.join(', '))
-    end
-    self
+  class OnestopIdBase
+
+    PREFIX = nil
+    MODEL = nil
+    NUM_COMPONENTS = 3
+
+    attr_accessor :geohash, :name
+
+    def initialize(string: nil, geohash: nil, name: nil, entity_prefix: nil)
+      if string && string.length > 0
+        @geohash = string.split(COMPONENT_SEPARATOR)[1]
+        @name = string.split(COMPONENT_SEPARATOR)[2]
+      elsif geohash && name
+        # Filter geohash and name; validate later
+        @geohash = geohash_filter(geohash)
+        @name = name_filter(name)
+      else
+        raise ArgumentError.new('either a string or entity/geohash/name must be specified')
+      end
+      # Check valid OnestopID
+      is_a_valid_onestop_id, errors = self.validate(self.to_s)
+      if !is_a_valid_onestop_id
+        raise ArgumentError.new(errors.join(', '))
+      end
+      self
+    end
+
+    def to_s
+      [self.class::PREFIX, @geohash, @name].join(COMPONENT_SEPARATOR)
+    end
+
+    def validate(value)
+      split = value.split(COMPONENT_SEPARATOR)
+      errors = []
+      errors << 'must not be empty' if value.blank?
+      errors << 'incorrect length' unless split.size == self.class::NUM_COMPONENTS
+      errors << 'invalid geohash' unless validate_geohash(split[1])
```

Comment:
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.